source/util/string_utils.h

```diff
@@ -56,23 +58,33 @@ inline std::vector<uint32_t> MakeVector(std::string input) {
         (byte_index < num_bytes ? uint8_t(input[byte_index]) : uint8_t(0));
     word |= (new_byte << (8 * (byte_index % sizeof(uint32_t))));
     if (3 == (byte_index % sizeof(uint32_t))) {
-      result.push_back(word);
+      result->push_back(word);
       word = 0;
     }
   }
   // Emit a trailing partial word.
   if ((num_bytes + 1) % sizeof(uint32_t)) {
-    result.push_back(word);
+    result->push_back(word);
   }
+}
+
+// Encodes a string as a sequence of words, using the SPIR-V encoding.
+inline std::vector<uint32_t> MakeVector(const std::string& input) {
+  std::vector<uint32_t> result;
+  AppendToVector(input, &result);
   return result;
 }
 
-// Decode a string from a sequence of words, using the SPIR-V encoding.
-template <class VectorType>
-inline std::string MakeString(const VectorType& words) {
+// Decode a string from a sequence of words between first and last, using the
+// SPIR-V encoding. Assert that a terminating 0-byte was found (unless
+// assertFoundTerminatingNull is passed as false).
+template <class InputIt>
+inline std::string MakeString(InputIt first, InputIt last,
+                              bool assertFoundTerminatingNull = true) {
   std::string result;
 
-  for (uint32_t word : words) {
+  for (InputIt pos = first; pos != last; ++pos) {
+    uint32_t word = *pos;
     for (int byte_index = 0; byte_index < 4; byte_index++) {
```

Comment:
I know you did not do this, but we should probably not be use "4" as a literal here.  We use 4 because we are assuming that "*pos" is 4 bytes.  We should provide a named constant for this, and add a static assert.

Reply:
Agreed, good catch.